experimental/framework/audience-heartbeat/src/audienceWithHeartBeat.ts

```diff
@@ -0,0 +1,102 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { EventEmitter } from "events";
+import { IAudience } from "@fluidframework/container-definitions";
+import { IClient, MessageType } from "@fluidframework/protocol-definitions";
+import { IInboundSignalMessage } from "@fluidframework/runtime-definitions";
+import { IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { IFluidAudienceWithHeartBeat } from "./interfaces";
+
+/**
+ * This class wraps implements heart-beat over Audience and emits event
+ * in case we are not listening from any client for 5 continuous heart-beats
+ * or we started getting heart-beat from a client who is not present in Audience.
+ */
+export class AudienceWithHeartBeat extends EventEmitter implements IFluidAudienceWithHeartBeat {
+    private readonly frequency: number;
+    private readonly audienceHeartBeat: Map<string, Date> = new Map();
+    private readonly audience: IAudience ;
+    private readonly runtime: IFluidDataStoreRuntime ;
+    private timer: any = undefined;
+
+    /**
+     * Creates a AudienceWithHeartBeat object.
+     * @param audience - Audience.
+     * @param frequency - heartbeat frequency in milliseconds.
+     */
+    constructor(
+        audience: IAudience,
+        runtime: IFluidDataStoreRuntime,
+        frequency: number = 30000) {
+        super();
+        audience.getMembers().forEach((client: IClient, clientId: string) => {
+            this.audienceHeartBeat.set(clientId, new Date());
+        });
+
+        this.audience = audience;
+        this.runtime = runtime;
+        this.frequency = frequency;
+    }
+
+    public get IFluidAudienceWithHeartBeat() {
+        return this;
+    }
+
+    /**
+     * {@inheritDoc (IFluidAudienceWithHeartBeat:interface).enableHeartBeat}
+     */
+    public enableHeartBeat() {
+        this.timer = setInterval(() => {
+            this.runtime.submitSignal("ping", {});
+            this.validateAudienceHeartBeat();
+        }, this.frequency);
+
+        // Listen for heartbeats
+        this.runtime.on("signal", (message: IInboundSignalMessage, local: boolean) => {
+            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+            if (this.timer !== undefined && message.clientId && message.type === "ping") {
+                this.audienceHeartBeat.set(message.clientId, new Date());
+
+                // client missed addMember event.
+                if (this.audience.getMember(message.clientId) === undefined) {
+                    this.emit(MessageType.ClientJoin, this.audience.getMember(message.clientId));
+                }
```

Comment:
Need to deal with the case where we miss a client ping for 5 times, (thus emitted client leave), but we heard for that client again.  (which I assume you want to emit join again?)

Reply:
I am emitting MessageType.ClientJoin and MessageType.ClientLeave both

Reply:
It's not clear to me how the API could be used and what the expectation is.  Can you write documentation and test to illustrate what can be built on top of this?